### ISSUE TITLE
Interface w/Same Name as Abstract Class Leads to Possibly Unexpected Behavior

### ISSUE THREAD

**Issue by aryeh-looker** [Microsoft/TypeScript#19347]
**TypeScript Version:**  `2.6.0-dev.20120171019`

**Code**

```ts
interface X {
    hello(name: String): void 
}

abstract class X { /* some optional stuff in here */ }

class Y extends X { }
```

**Expected behavior:**

Possibly a compiler error. Code like the following causes a compiler error:

```ts
abstract class A {
    abstract hello?(name: String): void
}

class B extends A { }
```

On the other hand, I can see an argument here that the code above (`class Y extends...`) should _not_ cause a compiler error because:

* we first define an interface
* we define a class X and extend its type
* the expression `class Y extends X` refers to the implementation of `X` and not necessarily its type, which says that `hello` should be implement

In other words, it might be better to implement the uppermost code like:

```ts
abstract class X { 
    abstract hello(name: String): void 
    /* some optional code here */
}

class Y extends X { } // will cause a compiler error
```

The code in the uppermost example is confusing, however, and _I think_ could incorrectly lead someone to believe it does something other than what the compiler. Possibly worth documenting if not fixing?

**Actual behavior:**

No compilation error.

**Comment by mhegazy** [comment]
Interfaces and classes merge in TS. please see https://www.typescriptlang.org/docs/handbook/declaration-merging.html for declaration merging in general.

by declaring `hello` on the interface `X`, you are telling the compiler there is a property that exists on this type that is not part of the class. one possibility is you will add it to the prototype explicitly, or you have a decorator that will always initialize it, etc.. 

The extends check on `abstract` classes is that the derived class has to implement *all* `abstract` members in the base class. `hello` here is not `abstract`.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
